iceoryx_posh/source/roudi/service_registry.cpp

```diff
@@ -16,45 +17,164 @@
 
 #include "iceoryx_posh/internal/roudi/service_registry.hpp"
 
+#include <algorithm>
+
 namespace iox
 {
 namespace roudi
 {
-void ServiceRegistry::add(const CaproIdString_t& service, const CaproIdString_t& instance)
+cxx::expected<ServiceRegistry::ServiceRegistryError>
+ServiceRegistry::add(const capro::ServiceDescription& serviceDescription)
 {
-    cxx::set::add(m_serviceMap[service].instanceSet, instance);
+    // Forbid duplicate service descriptions entries
+    for (auto& element : m_serviceDescriptionVector)
+    {
+        if (element == serviceDescription)
+        {
+            return cxx::error<ServiceRegistryError>(ServiceRegistryError::SERVICE_DESCRIPTION_ALREADY_ADDED);
+        }
+    }
+
+    if (!m_serviceDescriptionVector.push_back(serviceDescription))
+    {
+        return cxx::error<ServiceRegistryError>(ServiceRegistryError::SERVICE_REGISTRY_FULL);
+    }
+    m_serviceMap.insert({serviceDescription.getServiceIDString(), m_serviceDescriptionVector.size() - 1});
+    m_instanceMap.insert({serviceDescription.getInstanceIDString(), m_serviceDescriptionVector.size() - 1});
+    return cxx::success<>();
 }
 
-void ServiceRegistry::remove(const CaproIdString_t& service, const CaproIdString_t& instance)
+bool ServiceRegistry::remove(const capro::ServiceDescription& serviceDescription)
 {
-    cxx::set::remove(m_serviceMap[service].instanceSet, instance);
```

Comment:
Please use the suggested data structure: `std::map<std::map<CaproIdString_t, std::set<CaproIdString_t>>, CaproIdString_t> m_services` and refactor this method.

The implementation would be.
1. Remove event entry from `std::set<CaproIdString_t>`
2. If the set from 1. is empty remove the instance entry from `std::map<CaproIdString_t, std::set<CaproIdString_t>>`
3. If the map from 2. is empty remove the service entry from `m_services.

You can safely remove `m_serviceDescriptionVector`, `m_instanceMap` and `m_serviceMap` since the data structure from above is sufficient enough

Reply:
As described in another comment this structure is inefficient for storage (considering we would need static capacities for all of them) and also and does not lend itself for efficient operations, especially wildcard search. Furthermore the way it is proposed it it does not allow us to locate a full `ServiceDescription` object which will contain more than 3 strings in the foreseeable future - we will need to embed custom information of DDS topics I think (like QoS), have not fully thought it through. But even if we do not need it for QoS, it is good to have to attach all information about a service here in one place, or at least point to it. 

I am also not sure how a map as key would really behave, since it has no natural `operator<` which is a requirement for key types (would maybe fall back to addresses of the keys here, but this would be wrong for non-dynamic objects). Besides any map where keys or values are more than lightweight objects is doomed efficiency wise, since insertion and removal require copies of internal data (which would consist of maps of sets here). This is just not going to work efficiently

Reply:
See my answer in the [other thread](https://github.com/eclipse-iceoryx/iceoryx/pull/860#discussion_r655665071)

Reply:
To elaborate: Map keys need `operator<`, but if it does not exist it will use `std::less` in the map I think and use comparison on pointers to dynamically allocated data objects (allocated by the map). In other words: pointers to memory are the real key in this case, which is fine. Not fully sure though, our own map would have to behave in some reasonable way there as well and I do not know yet what this would be.

The storage size problem would be severe in any case for nested fixed-size containers.
This means we would need to implement a complex solution with a lot of drawbacks. We can have a simpler (still fairly complex) solution which at least avoids some of the drawbacks, is faster, generalizes in a better way to e.g. more keys, allows prefix based search and more.
The data structure itself is also somewhat easier to implement. But I fully intend to create the fixed-size map as well, just that we can design better algorithms use suitable abstractions instead for other problems.